docs:  migrate Express middleware from LB3 to LB4 (#4056)
strongloop/loopback-next: docs/site/migration/express-middleware.md

sidebar: lb4_sidebar
permalink: /doc/en/lb4/migration-express-middleware.html

> ---

Reviewer: Will there be docs here?

permalink: /doc/en/lb4/migration-express-middleware.html
---

> Migrating Express middleware from LoopBack 3 (LB3) application to LoopBack 4

Reviewer 2: There seems to be some duplicate content in this file.  Maybe we don't need the first 2 paragraphs?

I'm thinking if it's even better to only have the following content in this file: 
```
There are 2 steps involved in migrating a LoopBack 3 application with Express middleware:
1. ....
2. ....

See [this tutorial(..) for details. 
```

Or can we get the content of the https://github.com/strongloop/loopback-next/blob/master/examples/lb3-application/README.md here?

Author: Updated the links.

permalink: /doc/en/lb4/migration-express-middleware.html
---

This is a continuation of the
"[Mounting a LoopBack 3 application](https://loopback.io/doc/en/lb4/migration-mounting-lb3app.html)"
guide. The instructions will continue with the assumption that you have a base
application set up as instructed in
[this tutorial](https://github.com/strongloop/loopback-next/tree/master/examples/lb3-application#tutorial).

Now that you have a LoopBack 3 (LB3) app mounted on a LoopBack 4 (LB4) app, it
is time to further optimize the setup so that only the bare neccessary artifacts

> from the LoopBack 3 app remains, that includes moving all the middleware to a

Reviewer: nitpick: `remains, that includes` -> `remain. This includes`

{% include code-caption.html content="lb3app/server/config.json" %}

```json
{
  "restApiRoot": "/",
  "remoting": {
    "context": false,
    "rest": {
      "handleErrors": false,
      "handleUnknownPaths": false,
      "normalizeHttpPath": false,
      "xml": false
    },
    "json": {
      "strict": false,
      "limit": "100kb"
    },
    "urlencoded": {
      "extended": true,
      "limit": "100kb"
    },
    "cors": false
  }
}
```

We will be using a base Express app for mounting the LB4 app as described in
"[Creating an Express Application with LoopBack REST API](https://loopback.io/doc/en/lb4/express-with-lb4-rest-tutorial.html)"
guide.

Migrate the LB3 app's middleware from its `middleware.json` file to this Express
app, except the one from the `routes` phase (there is a
[pending task](https://github.com/strongloop/loopback-next/issues/4181) to
complete the support for this middleware).

{% include note.html content=" Each root property in the `middleware.json`
object represents a middleware phase, extract the relevant middleware and load
them in the Express app in order.

An entry like `"compression": {}` translates to `compression()`, and

> `loopback#favicon` translates to `loopback.favicon()` in plain JavaScript. For

Reviewer 3: ```suggestion
`loopback#favicon` translates to `loopback.favicon()` in TypeScript. For
```

}
```

Next, modify the `index.js` file in the root of the project to prevent the LB4
app from listening, by adding `listenOnStart: false` in `config.rest` object.
The `config` object should now look like this:

{% include code-caption.html content="index.js" %}

```js
const config = {
  rest: {
    port: +process.env.PORT || 3000,
    host: process.env.HOST || 'localhost',
    openApiSpec: {
      // useful when used with OpenAPI-to-GraphQL to locate your application
      setServersFromRequest: true,
    },
    listenOnStart: false,
  },
};
```

Then, in the `bootOptions` of the `CoffeeShopApplication` class, add the
`lb3app` to configure the path of the LB3 APIs.

{% include code-caption.html content="src/application.ts" %}

```js
lb3app: {
  mode: 'fullApp';
}
```

`this.bootOptions` should now look like this:

{% include code-caption.html content="src/application.ts" %}

```ts
this.bootOptions = {
  controllers: {
    // Customize ControllerBooter Conventions here
    dirs: ['controllers'],
    extensions: ['.controller.js'],
    nested: true,
  },
  lb3app: {
    mode: 'fullApp',
  },
};
```

Start the app:

```sh
$ npm start
```

> Load http://localhost:3000/ on your browser. This will load the Express app,

Reviewer 3: Please replace the naked URL with a markdown link.

```suggestion
Load [http://localhost:3000/](http://localhost:3000/) on your browser. This will load the Express app,
```

permalink: /doc/en/lb4/migration-express-middleware.html
---

This is a continuation of the
"[Mounting a LoopBack 3 application](https://loopback.io/doc/en/lb4/migration-mounting-lb3app.html)"
guide. The instructions will continue with the assumption that you have a base
app as instrcuted in
https://github.com/strongloop/loopback-next/tree/master/examples/lb3-application#tutorial.

Now that you have a LoopBack 3 (LB3) app mounted on a LoopBack 4 (LB4) app, it
is time to further optimize the setup so that only the bare neccessary artefacts

> from the LoopBack 3 app remains, that includes moving all the middleware to a

Author: Hmmm, softwares or software 😄?

> from the LoopBack 3 app remains, that includes moving all the middleware to a

Reviewer 3: I believe the plural of `middleware` is `middleware` again.

common location so that they are shared by both the LoopBack 3 and LoopBack 4
apps.

First off, edit the LB3 app's `config.json` file.

Remove these properties, they are no required anymore:

```json
"restApiRoot": "/api",
"host": "0.0.0.0",
"port": 3000,
```

> And then add `"handleUnknownPaths": false` to the `rest` property.

Reviewer: Can you explain why they should do this?

permalink: /doc/en/lb4/migration-express-middleware.html
---

This is a continuation of the
"[Mounting a LoopBack 3 application](https://loopback.io/doc/en/lb4/migration-mounting-lb3app.html)"
guide. The instructions will continue with the assumption that you have a base

> app as instrcuted in

Reviewer 3: Typo.

```suggestion
app as instructed in
```

> https://github.com/strongloop/loopback-next/tree/master/examples/lb3-application#tutorial.

Reviewer 3: Please replace the naked URL with a Markdown link.

common location so that they are shared by both the LoopBack 3 and LoopBack 4
apps.

First off, edit the LB3 app's `config.json` file.

Remove these properties, they are not required anymore:

```json
"restApiRoot": "/api",
"host": "0.0.0.0",
"port": 3000,
```

And then add `"handleUnknownPaths": false` to the `rest` property, this will
prevent the LB3 REST api from sending a 404 response for requests it cannot
handle.

The `config.json` file should now look like this:

{% include code-caption.html content="lb3app/server/config.json" %}

```json
{
  "remoting": {
    "context": false,
    "rest": {
      "handleErrors": false,
      "handleUnknownPaths": false,
      "normalizeHttpPath": false,
      "xml": false
    },
    "json": {
      "strict": false,
      "limit": "100kb"
    },
    "urlencoded": {
      "extended": true,
      "limit": "100kb"
    },
    "cors": false
  }
}
```

We will be using a base Express app for mounting the LB4 app. Notice how we are
loading the middleware from the LB3 in this app. These middleware will be shared
by both LB3 and LB4 apps. We are not adding the `loopback.rest` middleware, it

> will be loaded by the `Lb3AppBooterComponent`.

Reviewer 3: Please mention that these instructions are based on the approach described in https://loopback.io/doc/en/lb4/express-with-lb4-rest-tutorial.html, include a link to that doc page.

{% include code-caption.html content="lb3app/server.ts" %}

```ts
import {ApplicationConfig} from '@loopback/core';
import * as express from 'express';
import {Request, Response} from 'express';
import * as http from 'http';
import pEvent from 'p-event';
import * as path from 'path';

> import {CoffeeShopApplication} from './application';

Reviewer 3: People reading this guide will not have `CoffeeShopApplication`, please find a way how to make this example more generic, or at least add a comment mentioning that users have to use the name of _their_ application class instead of `CoffeeShopApplication`.

const loopback = require('loopback');
const compression = require('compression');
const cors = require('cors');
const helmet = require('helmet');

export class ExpressServer {
  private app: express.Application;
  public readonly lbApp: CoffeeShopApplication;
  private server?: http.Server;

  constructor(options: ApplicationConfig = {}) {
    this.app = express();
    this.lbApp = new CoffeeShopApplication(options);

>     // Middleware for LB3

Reviewer 3: ```suggestion
    // Middleware migrated from LoopBack 3
```

    this.app.use(loopback.favicon());
    this.app.use(compression());
    this.app.use(cors());

>     this.app.use(helmet());

Reviewer 3: I am concerned that this example shows only middleware from our CoffeeShop example application. Please find a way how to provide more generic instructions and explain users how to migrate `server/middleware.json` entries like `loopback#favicon` to TypeScript code using `this.app.use()`.

The code snippets should be examples illustrating the instructions written in plain English, not the instructions themselves.

Author: Added the detailed, starts at "Migrate the LB3 app's middleware from its ..."

  }

  public async boot() {
    await this.lbApp.boot();
  }

  public async start() {
    await this.lbApp.start();
    const port = this.lbApp.restServer.config.port || 3000;
    const host = this.lbApp.restServer.config.host || '127.0.0.1';
    this.server = this.app.listen(port, host);
    await pEvent(this.server, 'listening');
  }

  public async stop() {
    if (!this.server) return;
    await this.lbApp.stop();
    this.server.close();
    await pEvent(this.server, 'close');
    this.server = undefined;
  }
}
```

This app will replace the `CoffeeShopApplication` as the entry point for the
program, modify the `src/index.ts` file accordingly.

{% include code-caption.html content="src/index.ts" %}

```ts
import {ApplicationConfig} from '@loopback/core';
import {ExpressServer} from './server';

export {ExpressServer};

export async function main(options: ApplicationConfig = {}) {
  const server = new ExpressServer(options);
  await server.boot();
  await server.start();
  console.log('Server is running at http://127.0.0.1:3000');
}
```

Now create a static file named `express.html` to be served from the root of the

> server.

Reviewer 3: I have mixed feelings about this part. It makes sense to have Express/root-level public files in the context of an example showing how to mount LB4 API to an Express app. Such approach does not make so much sense to me in the context of migrating a LB3 application to LB4.

What I think will be a better approach: explain users how to take their public files in LB3 app (typically `lb3app/client`) and expose them either via the top-level Express app or via LB4.

Ideally, the instructions should be generic and explain how to migrate any public files (mounted via `express.static`), not only the default LB3 setup (`client` directory in the LB3 project).

Author: Added "Move any static files from the LB3 app to the `public` directory of the Express
app.", and removed irrelevant instruction.

      LoopBack main page and <a href="/hello/">here</a> to say hi.
    </h4>
  </body>
</html>
```

Then, update the `public/index.html` file to reflect the latest changes in the
path of the API. `/openapi.json` will become `/api/openapi.json`, and
`/explorer` will become `/api/explorer`.

Next, modify the `index.js` file in the root of the project to prevent the LB4
app from listening, by adding `listenOnStart: false` in `config.rest` object.
The `config` object should now look like this:

{% include code-caption.html content="index.js" %}

```js
const config = {
  rest: {
    port: +process.env.PORT || 3000,
    host: process.env.HOST || 'localhost',
    openApiSpec: {
      // useful when used with OpenAPI-to-GraphQL to locate your application
      setServersFromRequest: true,
    },
    listenOnStart: false,
  },
};
```

Then, in the `bootOptions` of the `CoffeeShopApplication` class, add the
`lb3app` to configure the path of the LB3 APIs.

{% include code-caption.html content="src/application.ts" %}

```js
lb3app: {
  // only REST routes are mounted
  mode: 'restRouter',
  restApiRoot: '/',
}
```

{% include note.html content="

> `restApiRoot` determines where the LB3 api will be mounted, if the value is `/`, it will be mounted on `/api`. If you want to mount it on a separate path you can set the value to something like `/lb3`, then the api will be mounted on `/api/lb3`.

Reviewer 3: :+1:

permalink: /doc/en/lb4/migration-express-middleware.html
---

Migrating app-level Express middleware from LoopBack 3 (LB3) to LoopBack 4 (LB4)

> requires a wrapper Express app, which will the mount the LB4 and also act as the

Reviewer 4: which will the mount the LB4 -> `which mounts LB4..`

point for mounting the shared Express middleware.

First, make sure you have done the necessary steps to mount your LB3 app on a
LB4 app as described in
[this tutorial](https://github.com/strongloop/loopback-next/tree/master/examples/lb3-application#tutorial).

> Then, create a directory which will be the wrapper Express app. Move the LB4 app

Reviewer 4: Do we want to mention the directory `migrate-express-middelware/` that we use as an example to make it more clear?

into this directory, and move the `package.json` file from the LB4 app to this

> directory.

Reviewer 3: I find this proposal problematic for several reasons.

1. It's not consistent with the solution we are describing in [Creating an Express Application with LoopBack REST API](https://loopback.io/doc/en/lb4/express-with-lb4-rest-tutorial.html#create-a-facade-express-application). This inconsistency may create confusion among our users, and also make support more time consuming for us - we will have to learn which Express-mounting approach is the developer using.

2. I find the changes too involved. Mounting LB4 app in a top-level Express application is a short-term workaround until we implement first-class support (see https://github.com/strongloop/loopback-next/issues/1293). Once that feature is landed, we will want the user to migrate away from top-level Express app and use LB4 as the top-level container. Ideally, such change should be mostly about removing code that's no longer needed, it should not require moving pretty much all LB files to a different location!

3. Finally, `lb4` CLI expects LoopBack files to be in the project root, alongside with `package.json`. This is important e.g. when adding a new datasource, where we want to add the connector to `package.json` dependencies.

Personally, I'd like the migration guide to follow the project layout described in our existing docs, where the LoopBack application stays in the project root directory and Express-specific things are implemented in newly added source files.

Reviewer: Based on the issue description, the layout should resemble:
```
express app
 - Express middleware
 - lb4-app
   - lb3-app
```
But from this comment:
> Personally, I'd like the migration guide to follow the project layout described in our existing docs, where the LoopBack application stays in the project root directory and Express-specific things are implemented in newly added source files.

It seems as though it should be
```
lb4-app
 - Express middleware
 - lb3-app
```

Can you clarify if this is what you mean?

But I do agree the `package.json` should remain at the lb4 app level.

Reviewer 3: Sorry for the confusion. The layout description is describing composition of Express/LB3/LB4 blocks at runtime.

The layout of project files should follow the usual LB4 project conventions, as described in _ Creating an Express Application with LoopBack REST API_.

> Update the `build` script to `"cd lb4-app && npx lb-tsc"`, this will enable us

Reviewer 4: I think we need to be more specific here. 

Maybe something like: 
Update the `build` script of the `package.json` file from ... to ... . This enables us...

to build the LB4 app from the root of the wrapper directory.

Install the dependencies in the wrapper directory by running `npm i`.

Next, we have to create an `index.js` file in the wrapper directory, which will
mount the LB4 app.

> {% include code-caption.html content="index.js" %}

Reviewer 3: Let's not mix JavaScript and TypeScript in a single project please, that's confusing. Please use TypeScript everywhere.

```js
const express = require('express');
const http = require('http');
const path = require('path');
const pEvent = require('p-event');

const Lb4AppApplication = require('./lb4-app/dist/application')
  .Lb4AppApplication;

const app = express();
const config = {
  rest: {
    port: +process.env.PORT || 3000,
    host: process.env.HOST || 'localhost',
    openApiSpec: {
      // useful when used with OpenAPI-to-GraphQL to locate your application
      setServersFromRequest: true,
    },
    // Use the LB4 application as a route. It should not be listening.
    listenOnStart: false,
  },
};
const lbApp = new Lb4AppApplication(config);

app.use('/api', lbApp.requestHandler);

// Express middleware shared by LB4 and LB3 apps
app.get('/hello', function(req, res) {
  res.send('Hello world!');

> });

Reviewer 3: To me, this seems unrelated to the migration path. Can you remove this part please?

app.use(express.static(path.join(__dirname, '../public')));

async function start() {
  await lbApp.boot();
  await lbApp.start();
  const port = lbApp.restServer.config.port;
  const host = lbApp.restServer.config.host;
  const server = app.listen(port, host);
  await pEvent(server, 'listening');
  console.log(`Server is running at http://${host}:${port}`);
}

> start();

Reviewer: `await start();`?

```

This file acts as the entry point for starting the LB4 app and also provides the
familiar Express APIs via the `app` Express instance.

> 

Reviewer 3: I think you forgot to write instructions for migrating middleware defined in LB3 `server/middleware.json` to the Express app file, PTAL.

Also, once the middleware is registered outside of LB3, we should switch from `fullApp` mode to `restRouter` mode, see  https://loopback.io/doc/en/lb4/migration-mounting-lb3app.html#options and https://github.com/strongloop/loopback-next/blob/6a77bb76f58506b2d5beab65ed9e13ba2f58f3e7/packages/booter-lb3app/src/lb3app.booter.ts#L145-L150

Make sure to run `npm run build` to compile the LB4 app written in TyepScript to
JavaScript any time you make changes to the app.

Start the app as usual using `npm start`.

A working example of migrating Express middleware from LB3 to LB4 app can be
found at

> https://github.com/strongloop/loopback-next/tree/master/examples/migrate-express-middleware.

Reviewer 3: I have mixed feelings about introducing a new example app for this migration step. While adding an example app is easy, the long-term maintenance costs are non-negligible. Have you considered updating `examples/lb3-application` instead?

@strongloop/loopback-maintainers I'd like to get more opinions in this aspect. WDYT?

Reviewer: +1 to updating the existing lb3 application. This feels like the next step for it since I assume the application is meant to eventually show the full migration path.

Reviewer 2: +1 in reusing the `examples/lb3-application` if appropriate.
Would we be adding more examples specifically for migration?

Author: The PR is being redone using `examples/lb3-application` as the example app. There was a confusion regarding the structure of the app, it's cleared now.

Reviewer 5: + for `examples/lb3-application`. @hacksparrow could you also update the tests for the example app?
The updated routes seem not working, see https://travis-ci.com/strongloop/loopback-next/jobs/256165615

Author: Thanks @agnes512 

@bajtos this looks like a config issue somewhere. Are you able to point it out in a quick glance? I am working on it, though.
